### PR TITLE
Download file API in c8y-http-proxy actor

### PR DESCRIPTION
**Pull request by albinsuresh**
## Proposed changes

Implement download file API in c8y-http-proxy actor which is used by c8y-config-plugin to download config file updates

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

- [ ] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [ ] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [ ] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: crates/bin/c8y_configuration_poc/src/c8y_http_proxy/actor.rs [hunk 1]
```diff
@@ -293,8 +325,33 @@ impl C8YHttpProxyActor {
...
+            Ok(token)
+        } else {
+            Err(C8YRestError::CustomError("JWT token not available".into()))
+        }
+    }
+
+    async fn download_file(&mut self, request: DownloadFile) -> Result<Unit, C8YRestError> {
+        let mut download_info: DownloadInfo = request.download_url.as_str().into();
+        // If the provided url is c8y, add auth
+        if self.url_is_in_my_tenant_domain(download_info.url()) {
```
Comment (didier-wenzek) [inline crates/bin/c8y_configuration_poc/src/c8y_http_proxy/actor.rs:339]:
```suggestion
        if self.end_point.url_is_in_my_tenant_domain(download_info.url()) {
```

File: crates/bin/c8y_configuration_poc/src/c8y_http_proxy/actor.rs [hunk 2]
```diff
@@ -162,6 +169,31 @@ impl C8YHttpProxyActor {
         info!(target: self.name(), "initialisation done.");
     }
 
+    pub fn url_is_in_my_tenant_domain(&self, url: &str) -> bool {
```
Comment (didier-wenzek) [inline crates/bin/c8y_configuration_poc/src/c8y_http_proxy/actor.rs:172]: It's useless to have this method as the `self` owns a `C8yEndPoint` that provides this method.

File: crates/bin/c8y_configuration_poc/src/c8y_http_proxy/mod.rs [hunk 3]
```diff
@@ -34,6 +37,7 @@ mod tests;
 pub struct C8YHttpConfig {
     pub c8y_host: String,
     pub device_id: String,
+    pub tmp_dir: PathBuf,
```
Comment (didier-wenzek) [inline crates/bin/c8y_configuration_poc/src/c8y_http_proxy/mod.rs:40]: Why the proxy would need a temporary directory? To temporarily download files there? Cannot this information passed to the download method?

File: crates/common/download/Cargo.toml [hunk 4]
```diff
@@ -11,10 +11,13 @@ description = "download_manager"
...
 log = "0.4"
 nix = "0.24"
-reqwest = { version = "0.11", default-features = false, features = ["rustls-tls"] }
+reqwest = { version = "0.11", default-features = false, features = [
+    "rustls-tls",
+] }
 serde = { version = "1.0", features = ["derive"] }
 serde_json = "1.0"
 thiserror = "1.0"
+tedge_utils = { path = "../tedge_utils" }
```
Comment (didier-wenzek) [inline crates/common/download/Cargo.toml:20]: What changed to have a new dependency?

Okay, I see that you added some permission checks and renaming features. But this seems unrelated to this PR about actors. I you think these are important improvements, then a specific PR would be better.

File: crates/bin/c8y_configuration_poc/src/c8y_http_proxy/mod.rs (hunk 3 above)
Comment (albinsuresh) [reply crates/bin/c8y_configuration_poc/src/c8y_http_proxy/mod.rs:40]: Yeah i wasn't quite happy with this either and updating the `download` method signature is another option as well, sure. This was done to keep the existing `Download` API happy, even though I really wanted to change that API, but decided against it to minimize side effects from existing users of it.

File: crates/common/download/Cargo.toml (hunk 4 above)
Comment (albinsuresh) [reply crates/common/download/Cargo.toml:20]: Actually now I'm more convinced that all that permission check stuff done in the `rename` API should stay in the `tedge_util` crate rather than the `download` crate. So, I'll move that code there and remove this dependency.

File: crates/bin/c8y_configuration_poc/src/c8y_http_proxy/mod.rs (hunk 3 above)
Comment (albinsuresh) [reply crates/bin/c8y_configuration_poc/src/c8y_http_proxy/mod.rs:35]: It's unavoidable in the current state as the proxy is the one creating and invoking the downloader now. So, it needs to know where the `tmp_dir` is.

File: crates/common/download/src/download.rs [hunk 5]
```diff
@@ -82,6 +91,15 @@ impl Downloader {
     }
 
     pub async fn download(&self, url: &DownloadInfo) -> Result<(), DownloadError> {
```
Comment (didier-wenzek) [inline crates/common/download/src/download.rs:93]: Beware that there is another PR aiming to refactor the download API: https://github.com/thin-edge/thin-edge.io/pull/1709

File: crates/core/tedge_actors/src/errors.rs [hunk 6]
```diff
@@ -1,17 +1,20 @@
...
+#[derive(Error, Debug)]
 pub enum ChannelError {
     #[error("Fail to send a message: the receiver has been dropped")]
     SendError(#[from] futures::channel::mpsc::SendError),
 
     #[error("Fail to receive a message: the sender has been dropped")]
     ReceiveError(),
+
+    #[error(transparent)]
+    ActorError(#[from] anyhow::Error),
```
Comment (didier-wenzek) [inline crates/core/tedge_actors/src/errors.rs:13]: This is more related to a `RuntimeError`. But for the latter to work, one has to change the  type signature of the `Actor::run()` method to return `RuntimeError`. This would makes more sense.

**Review by didier-wenzek — APPROVED** [review]
Approved
